In [3]:
import pandas as pd

brewers = pd.read_csv('data/brewers_batting_2025.csv')
brewers

,Rk,Player,Age,Pos,WAR,G,PA,AB,R,H,...,Rbat+,TB,GIDP,HBP,SH,SF,IBB,Pos.1,Awards,Player-additional
0,1.0,William Contreras,27.0,C,3.8,150,659,566,89,147,...,116,226,14,3,0,6,4,*2D/37,NaN,contrwi02
1,2.0,Rhys Hoskins,32.0,1B,0.8,90,328,279,30,66,...,107,116,6,5,0,6,2,3/HD,NaN,hoskirh01
2,3.0,Brice Turang*,25.0,2B,5.5,156,659,584,97,168,...,122,254,3,2,2,5,1,*4/H6,MVP-14,turanbr02
3,4.0,Joey Ortiz,26.0,SS,0.2,149,506,470,62,108,...,61,149,10,3,6,0,0,*6/H,NaN,ortizjo06
4,5.0,Caleb Durbin,25.0,3B,2.7,136,506,445,60,114,...,105,172,10,24,3,4,0,*54H/6,ROY-3,durbica01
5,6.0,Isaac Collins#,27.0,LF,2.0,130,441,372,56,98,...,124,153,8,6,3,3,0,*7H/9D54,ROY-4,colliis01
6,7.0,Jackson Chourio,21.0,CF,2.2,131,589,549,88,148,...,111,254,13,3,1,6,1,*879/D,NaN,chourja01
7,8.0,Sal Frelick*,25.0,RF,2.9,142,594,528,76,152,...,109,214,4,7,3,5,0,*98/H35,NaN,frelisa01
8,9.0,Christian Yelich*,33.0,DH,3.0,150,644,573,88,151,...,120,259,6,6,0,1,6,*D7/H,MVP-12,yelicch01
9,10.0,Andrew Vaughn,27.0,1B,1.3,64,254,221,26,68,...,138,109,8,3,1,5,0,3/H,NaN,vaughan01


In [7]:
import pandas as pd

teams = pd.read_csv('data/team_batting_2025.csv')
print(teams.shape)
teams

(32, 29)


,Tm,#Bat,BatAge,R/G,G,PA,AB,R,H,2B,...,SLG,OPS,OPS+,TB,GDP,HBP,SH,SF,IBB,LOB
0,Arizona Diamondbacks,65,27.8,4.88,162,6210,5480,791,1377,277,...,0.433,0.757,109,2372,109,81,37,64,21,1112
1,Athletics,58,26.1,4.52,162,6151,5547,733,1403,296,...,0.431,0.749,104,2388,117,44,18,35,14,1106
2,Atlanta Braves,71,28.3,4.47,162,6186,5508,724,1349,243,...,0.399,0.720,102,2200,98,51,15,36,14,1160
3,Baltimore Orioles,70,26.5,4.18,162,6020,5416,677,1273,251,...,0.394,0.699,95,2135,102,75,4,41,14,1044
4,Boston Red Sox,56,27.6,4.85,162,6206,5562,786,1414,324,...,0.421,0.745,108,2344,99,72,13,41,24,1125
5,Chicago Cubs,52,28.4,4.90,162,6162,5495,793,1371,267,...,0.430,0.751,114,2365,72,46,10,55,16,1108
6,Chicago White Sox,62,26.8,3.99,162,5987,5377,647,1250,243,...,0.373,0.675,88,2008,108,49,22,36,8,1055
7,Cincinnati Reds,53,27.2,4.42,162,6082,5448,716,1333,250,...,0.391,0.706,92,2130,106,51,12,44,15,1088
8,Cleveland Guardians,45,27.4,3.97,162,5943,5310,643,1199,243,...,0.373,0.670,86,1982,78,60,28,50,30,1016
9,Colorado Rockies,57,26.7,3.69,162,5908,5408,597,1281,253,...,0.386,0.679,80,2088,118,48,21,35,5,986


In [8]:
brewers_row = teams[teams['Tm'] == 'Milwaukee Brewers']
league_avg = teams[teams['Tm'] == 'League Average']

brewers_row

,Tm,#Bat,BatAge,R/G,G,PA,AB,R,H,2B,...,SLG,OPS,OPS+,TB,GDP,HBP,SH,SF,IBB,LOB
16,Milwaukee Brewers,55,26.9,4.98,162,6227,5510,806,1423,265,...,0.403,0.736,103,2222,97,73,26,50,15,1158


In [9]:
compare_cols = ['R', 'HR', '2B', '3B', 'SB', 'BB', 'SO', 'BA', 'OBP', 'SLG']

comparison = pd.concat([
    brewers_row[compare_cols].reset_index(drop=True),
    league_avg[compare_cols].reset_index(drop=True)
])
comparison.index = ['Brewers', 'League Average']
comparison

,R,HR,2B,3B,SB,BB,SO,BA,OBP,SLG
Brewers,806,166,265,18,164,564,1266,0.258,0.332,0.403
League Average,720,188,258,21,115,513,1355,0.245,0.315,0.404


Roster-Level Gap Analysis: Offesive Power

**Finding:** The Brewers offense is contact-oriented rather than power-oriented, which may be a roster weakness against strong pitching

**Supporting Data (2025 season, Brewers vs. league average):** 
- Batting Average: .258 vs .245 (+5%)
- On-Base Percentage: .332 vs .315 (+5%)
- Slugging Percentage: .403 vs .404 (essentially even)
- Home Runs: 166 vs 188 (below average)
- Strikeouts: 1266 vs 1355 (well below average)

**Interpretation:** Despite an above-average ability to get on base and make contact, Milwauke gets almiost no extra production from power. Their slugging is only league-average despite a higher batting average, suggesting they're compensating with doubles/triples rather than home runs. This profile can be exploited by strong pitching, where contact alone is less effective than raw power is producing runs


In [11]:
pitching = pd.read_csv('data/team_pitching_2025.csv')
print(pitching.shape)
pitching

(32, 36)


,Tm,#P,PAge,RA/G,W,L,W-L%,ERA,G,GS,...,BF,ERA+,FIP,WHIP,H9,HR9,BB9,SO9,SO/W,LOB
0,Arizona Diamondbacks,43,29.8,4.85,80,82,0.494,4.49,162,162,...,6183,96,4.27,1.318,8.7,1.2,3.1,8.0,2.58,1093
1,Athletics,36,28.6,5.04,76,86,0.469,4.70,162,162,...,6232,92,4.66,1.361,8.7,1.4,3.6,8.3,2.32,1120
2,Atlanta Braves,46,29.4,4.53,76,86,0.469,4.36,162,162,...,6116,96,4.20,1.306,8.4,1.2,3.3,8.9,2.67,1088
3,Baltimore Orioles,42,30.6,4.86,75,87,0.463,4.60,162,162,...,6166,88,4.41,1.365,9.0,1.4,3.3,8.5,2.58,1098
4,Boston Red Sox,36,29.4,4.17,89,73,0.549,3.70,162,162,...,6150,109,3.98,1.286,8.3,1.0,3.3,8.5,2.57,1156
5,Chicago Cubs,34,31.0,4.01,92,70,0.568,3.79,162,162,...,5904,104,4.16,1.177,8.1,1.3,2.5,7.9,3.12,965
6,Chicago White Sox,33,27.5,4.58,60,102,0.370,4.26,162,162,...,6124,98,4.47,1.364,8.5,1.2,3.8,8.2,2.16,1148
7,Cincinnati Reds,29,29.1,4.20,83,79,0.512,3.86,162,162,...,6035,116,4.11,1.222,7.9,1.2,3.1,8.7,2.79,1070
8,Cleveland Guardians,26,27.1,4.01,88,74,0.543,3.70,162,162,...,6054,111,3.95,1.261,8.1,1.1,3.3,8.6,2.62,1098
9,Colorado Rockies,31,27.7,6.30,43,119,0.265,5.97,162,162,...,6367,82,5.24,1.582,10.7,1.6,3.5,7.0,1.97,1135


In [12]:
brewers_pitch = pitching[pitching['Tm'] == 'Milwaukee Brewers']
brewers_pitch

,Tm,#P,PAge,RA/G,W,L,W-L%,ERA,G,GS,...,BF,ERA+,FIP,WHIP,H9,HR9,BB9,SO9,SO/W,LOB
16,Milwaukee Brewers,33,28.4,3.91,97,65,0.599,3.58,162,162,...,6041,119,3.91,1.228,7.7,1.0,3.3,8.9,2.68,1098


In [13]:
league_avg_pitch = pitching[pitching['Tm'] == 'League Average']
league_avg_pitch['ERA']

30    4.15
Name: ERA, dtype: float64

In [14]:
brewers_iso = brewers_row['SLG'].values[0] - brewers_row['BA'].values[0]
league_iso = league_avg['SLG'].values[0] - league_avg['BA'].values[0]

print(f"Brewers ISO: {brewers_iso:.3f}")
print(f"League Average ISO: {league_iso:.3f}")

Brewers ISO: 0.145
League Average ISO: 0.159


In [18]:
statcast = pd.read_csv('data/statcast_batters_2025.csv')
print(statcast.shape)
statcast.columns.tolist()

(251, 18)


['last_name, first_name',
 'player_id',
 'attempts',
 'avg_hit_angle',
 'anglesweetspotpercent',
 'max_hit_speed',
 'avg_hit_speed',
 'ev50',
 'fbld',
 'gb',
 'max_distance',
 'avg_distance',
 'avg_hr_distance',
 'ev95plus',
 'ev95percent',
 'barrels',
 'brl_percent',
 'brl_pa']

In [19]:
statcast = pd.read_csv('data/statcast_batters_2025.csv')
print(statcast.shape)
statcast.columns.tolist()

(251, 18)


['last_name, first_name',
 'player_id',
 'attempts',
 'avg_hit_angle',
 'anglesweetspotpercent',
 'max_hit_speed',
 'avg_hit_speed',
 'ev50',
 'fbld',
 'gb',
 'max_distance',
 'avg_distance',
 'avg_hr_distance',
 'ev95plus',
 'ev95percent',
 'barrels',
 'brl_percent',
 'brl_pa']

In [20]:
print(f"League Avg Barrel%: {statcast['brl_percent'].mean():.1f}")
print(f"League Avg HardHit%: {statcast['ev95percent'].mean():.1f}")

League Avg Barrel%: 9.2
League Avg HardHit%: 42.1


In [21]:
print(f"League Avg Barrel%: {statcast['brl_percent'].mean():.1f}")
print(f"League Avg HardHit%: {statcast['ev95percent'].mean():.1f}")

League Avg Barrel%: 9.2
League Avg HardHit%: 42.1


In [22]:
candidates = ['Alonso, Pete', 'Santander, Anthony', 'Grisham, Trent', 'Grichuk, Randal', "O'Neill, Tyler"]
statcast[statcast['last_name, first_name'].isin(candidates)][['last_name, first_name', 'brl_percent', 'ev95percent']]

,"last_name, first_name",brl_percent,ev95percent
27,"Alonso, Pete",18.9,54.4
131,"Grisham, Trent",14.2,46.4


In [23]:
statcast[statcast['last_name, first_name'].str.contains('Santander|Grichuk|Neill', case=False, na=False)]

,"last_name, first_name",player_id,attempts,avg_hit_angle,anglesweetspotpercent,max_hit_speed,avg_hit_speed,ev50,fbld,gb,max_distance,avg_distance,avg_hr_distance,ev95plus,ev95percent,barrels,brl_percent,brl_pa
